File: java/8_StringtoInteger(atoi).java
class Solution {
    public int myAtoi(String str) {
        if (str == null || str.length() == 0) return 0;
        int res = 0;
        boolean flag = true;
        char[] s = str.toCharArray();
        int i = 0;
        while (s[i] == ' ' &&  (i+1) < s.length) {
            i++;
        }

        if (s[i] == '-') flag = false;
        if (s[i] == '-' || s[i] == '+') i++;

        while (i < s.length && isDigit(s[i])) {
            int r = s[i] - '0';
            if (res > Integer.MAX_VALUE / 10 || (res == Integer.MAX_VALUE / 10 && r > 7)) {
                return flag ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            res = res * 10 + r;
            i++;
        }

        return flag ? res : -res;
    }

    private boolean isDigit(char c) {
        if (c >= '0' && c <= '9') return true;
        else return false;
    }
}

File: c++/8_StringtoInteger(atoi).cpp
class Solution {
public:
    int myAtoi(string s) {
        int k = 0;
        while (k < s.size() && s[k] == ' ') k++;
        if (k == s.size()) return 0;
        
        bool flag = true;
        if (s[k] == '-') flag = false, k++;
        else if (s[k] == '+') k++;
        
        int res = 0;
        while (k < s.size() && s[k] >= '0' && s[k] <= '9') {
            int x = s[k] - '0';
            if (flag && res > (INT_MAX - x) / 10) return INT_MAX;
            if (!flag && -res < (INT_MIN + x) / 10) return INT_MIN;
            if (-res * 10 - x == INT_MIN) return INT_MIN;
            res = res * 10 + x;
            k++;
        }
        return flag ? res : -res;
    }
};
class Solution {
public:
    int myAtoi(string s) {
        int k = 0;
        while (k < s.size() && s[k] == ' ') k++;

        bool flag = true;
        if (s[k] == '-') flag = false, k++;
        else if (s[k] == '+') k++;
        long long res = 0;
        while (k < s.size() && s[k] >= '0' && s[k] <= '9') {
            res = res * 10 + s[k] - '0';
            k++;
            if (res > INT_MAX) break;
        }

        res = flag ? res : -res;
        if (res > INT_MAX) res = INT_MAX;
        if (res < INT_MIN) res = INT_MIN;
        return res;
    }
};